_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,571 @@
+pull-request contributes (or removes!) from the total. Now that we most of the
+data we need, this is possible!
+
+### Code Size Deltas on Pull Requests
+
+To get deltas on pull requests, we'll need CI to run a script whenever one is
+published or updated. This script will take a local binary, compute its code
+size, and compare it against a given revision or the
+`git-merge-base`[^git_merge_base] of the "master" branch. `git-merge-base` takes
+both revisions given and finds the most recent common ancestor between the two.
+In the case of the Zephyr Project, this is typically finding the parent commit
+between the base of a pull-request and the `origin/master` branch.
+
+In our
+[example code](https://github.com/memfault/interrupt/tree/master/example/tracking-code-size/codesizes.py),
+we have a command, `diff` which will do exactly this.
+
+```python
+@cli.command()
+@click.argument("binary")
+@click.option("--build", default="default")
+@click.option("--revision")
+def diff(binary, build, revision):
+    """
+    Calculate the size of the BINARY given, and compare it
+    to either the REVISION provided or the "merge-base" of the
+    current branch and master for the given BUILD.
+    """
+    text, data, bss = _calculate_codesize(binary)
+    if not revision:
+        click.echo("Using `git merge-base` against master branch")
+        # git merge-base HEAD master
+        revision = _get_merge_base("HEAD", "master")
+
+    codesize = _fetch_codesize(build, revision)
+    if not codesize:
+        click.echo("No code size for build: {}, revision: {}".format(
+            build, revision))
+        return
+
+    text_delta = text - codesize.text
+    data_delta = data - codesize.data
+    bss_delta = bss - codesize.bss
+
+    click.echo("Comparing current binary against: {}".format(revision))
+    click.echo(dict(text=text_delta, data=data_delta, bss=bss_delta))
+```
+
+Let's try it out! I went digging into the Zephyr Github Pull Requests page for a
+large changeset, and found a good candidate:
+[Timeout API rework #23382](https://github.com/zephyrproject-rtos/zephyr/pull/23382).
+
+We'll check out the branch and build the firmware first.
+
+```
+# Check out the branch for #23382
+$ cd zephyrproject/zephyr
+$ git remote add pr git@github.com:andyross/zephyr.git
+$ git fetch pr
+$ git checkout pr/timeout-take-2
+
+# Build the firmware
+$ west build -b cc3220sf_launchxl samples/net/lwm2m_client --pristine
+```
+
+Next, let's sanity check and ensure that the pull-request is based against a
+commit in the last 2,500, which would mean we had uploaded the size.
+
+```
+$ git merge-base master HEAD
+7c5db4b755f73ab5f93a6456abaa5b97bd6aff7a
+
+$ python ~/codesizes.py fetch 7c5db4b755f73ab5f93a6456abaa5b97bd6aff7a --build lwm2m_client
+{'text': 146334, 'data': 4220, 'bss': 41735}
+```
+
+Perfect! We have the commit's code size in our database. Let's get the delta.
+
+```
+$ python ~/codesizes.py diff build/zephyr/zephyr.elf --build lwm2m_client
```

Comment:
hrm I don't seem to be seeing this result. Any chance you were using a different version of gcc on aws ec2 instance vs locally?

```
# arm-none-eabi-size zephyr_before.elf
   text	   data	    bss	    dec	    hex	filename
 146336	   4220	  41735	 192291	  2ef23	zephyr_before.elf
(.pyocd_venv) chrisc: ~/dev/int_zephyr/zephyr/

# arm-none-eabi-size zephyr_after.elf
   text	   data	    bss	    dec	    hex	filename
 146584	   4220	  41735	 192539	  2f01b	zephyr_after.elf
```

Reply:
Fixed! Thanks for calling this out.